### PR TITLE
Adjust delegate caching logic for abstract statics.

### LINKED ISSUE #60069: Static abstract delegate creation within generic code cached into non-generic location
**Version Used**: 4.2.0-3.22159.2

**Steps to Reproduce**:

```csharp
using System;

class TestStaticInterfaceMethods
{
    interface ISimple
    {
        static abstract string GetCookie();
    }

    class SimpleClass : ISimple
    {
        public static string GetCookie() => "SimpleClass";
    }

    struct SimpleStruct : ISimple
    {
        public static string GetCookie() => "SimpleStruct";
    }

    static void TestSimpleInterface<T>(string expected) where T : ISimple
    {
        Func<string> d = T.GetCookie;
        string actual = d();
        if (actual != expected)
        {
            throw new Exception($"{actual} != {expected}");
        }
    }

    static void Main()
    {
        TestSimpleInterface<SimpleClass>("SimpleClass");
        TestSimpleInterface<SimpleStruct>("SimpleStruct");
    }
}
```

**Expected Behavior**:
The above program finishes without exception.

**Actual Behavior**:
The above program throws `'SimpleClass != SimpleStruct'`.

This is because the delegate creation gets cached into a static field on a non-generic type and the second invocation of `TestSimpleInterface` ends up reusing the delegate from previous invocation.

```
.class nested private auto ansi abstract sealed beforefieldinit '<>O'
	extends [System.Runtime]System.Object
{
	.custom instance void [System.Runtime]System.Runtime.CompilerServices.CompilerGeneratedAttribute::.ctor() = (
		01 00 00 00
	)
	// Fields
	.field public static class [System.Runtime]System.Func`1<string> '<0>__GetCookie'

} // end of class <>O
```

Above is how Roslyn parks the cached delegate. Notice the class is not generic.

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #60069.

### CODE REVIEW

**Review by 333fred — APPROVED** [review]

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For the second review.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For the second review.

File: src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs [hunk 1]
```diff
@@ -12309,6 +12309,132 @@ class Test
...
+        Func<string> d2 = TestSimpleInterface<SimpleStruct>();
+
+        System.Console.WriteLine(d1());
+        System.Console.WriteLine(d2());
+
+        for (int i = 0; i < 5; i++)
+        {
+            if (d1 != (object)TestSimpleInterface<SimpleClass>())    
+            {
+                System.Console.WriteLine(""d1 is not cheched!!!"");
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs:12339]: nit: spelling was probably meant to be 'cached' here and below.

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs (hunk 1 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs:12339]:
```suggestion
                System.Console.WriteLine(""d1 is not cached!!!"");
```

File: src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs [hunk 2]
```diff
@@ -12309,6 +12309,132 @@ class Test
...
+            }
+
+            if (d2 != (object)TestSimpleInterface<SimpleStruct>())    
+            {
+                System.Console.WriteLine(""d2 is not cheched!!!"");
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs:12344]:
```suggestion
                System.Console.WriteLine(""d2 is not cached!!!"");
```

File: src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs [hunk 3]
```diff
@@ -12309,6 +12309,132 @@ class Test
...
+
+        System.Console.WriteLine(d1());
+        System.Console.WriteLine(d2());
+
+        for (int i = 0; i < 5; i++)
+        {
+            (Func<string> d21, Func<string> d22) = TestSimpleInterface<SimpleClass, SimpleStruct>();
+            if (d1 != (object)d21)    
+            {
+                System.Console.WriteLine(""d1 is not cheched!!!"");
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs:12402]:
```suggestion
                System.Console.WriteLine(""d1 is not cached!!!"");
```

File: src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs [hunk 4]
```diff
@@ -12309,6 +12309,132 @@ class Test
...
+            }
+
+            if (d2 != (object)d22)    
+            {
+                System.Console.WriteLine(""d2 is not cheched!!!"");
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs:12407]:
```suggestion
                System.Console.WriteLine(""d2 is not cached!!!"");
```

File: src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs (hunk 1 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs:12339]:
```suggestion
                System.Console.WriteLine(""d1 is not cached!!!"");
```

File: src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs (hunk 2 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs:12344]:
```suggestion
                System.Console.WriteLine(""d2 is not cached!!!"");
```

File: src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs (hunk 3 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs:12402]:
```suggestion
                System.Console.WriteLine(""d1 is not cached!!!"");
```

File: src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs (hunk 4 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs:12407]:
```suggestion
                System.Console.WriteLine(""d2 is not cached!!!"");
```

**Pull request merged**